lib/sshutils/reply.go

```diff
@@ -0,0 +1,113 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package sshutils
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// Reply is a helper to handle replying/rejecting and log messages when needed.
+type Reply struct {
+	log *slog.Logger
+}
+
+// NewReply creates a new reply helper for SSH servers.
+func NewReply(log *slog.Logger) *Reply {
+	return &Reply{log: log}
+}
+
+// RejectChannel rejects the channel with provided message.
+func (r *Reply) RejectChannel(ctx context.Context, nch ssh.NewChannel, reason ssh.RejectionReason, msg string) {
+	err := nch.Reject(reason, msg)
+	if err != nil {
+		r.log.WarnContext(ctx, "Failed to reject channel", "error", err)
+	}
+}
+
+// RejectUnknownChannel rejects the channel with reason ssh.UnknownChannelType.
+func (r *Reply) RejectUnknownChannel(ctx context.Context, nch ssh.NewChannel) {
+	channelType := nch.ChannelType()
+	r.log.WarnContext(ctx, "Unknown channel type", "channel", channelType)
+	r.RejectChannel(ctx, nch, ssh.UnknownChannelType, fmt.Sprintf("unknown channel type: %v", channelType))
+}
+
+// RejectWithAcceptError rejects the channel when ssh.NewChannel.Accept fails.
+func (r *Reply) RejectWithAcceptError(ctx context.Context, nch ssh.NewChannel, err error) {
+	r.log.WarnContext(ctx, "Unable to accept channel", "channel", nch.ChannelType(), "error", err)
+	r.RejectChannel(ctx, nch, ssh.ConnectionFailed, fmt.Sprintf("unable to accept channel: %v", err))
+}
+
+// RejectWithNewRemoteSessionError rejects the channel when the corresponding
+// remote session fails to create.
+func (r *Reply) RejectWithNewRemoteSessionError(ctx context.Context, nch ssh.NewChannel, remoteError error) {
+	r.log.WarnContext(ctx, "Remote session open failed", "error", remoteError)
+	reason, msg := ssh.ConnectionFailed, fmt.Sprintf("remote session open failed: %v", remoteError)
+	var e *ssh.OpenChannelError
+	if errors.As(remoteError, &e) {
+		reason, msg = e.Reason, e.Message
+	}
+	r.RejectChannel(ctx, nch, reason, msg)
+}
+
+// ReplyError replies an error to an ssh.Request.
+func (r *Reply) ReplyError(ctx context.Context, ch ssh.Channel, req *ssh.Request, err error) {
+	r.log.ErrorContext(ctx, "failure handling SSH request", "request_type", req.Type, "error", err)
+	// Terminate the error with a newline when writing to remote channel's
+	// stderr so the output does not mix with the rest of the output if the remote
+	// side is not doing additional formatting for extended data.
+	// See github.com/gravitational/teleport/issues/4542
+	message := utils.FormatErrorWithNewline(err)
```

Comment:
Is git going to forward stderr to the user?

Reply:
good catch. i copied from existing impl without too much thought. it's not needed and removed now.